src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -1352,31 +1609,46 @@ internal ProcessInputWriter(InternalCommand command)
         {
             Dbg.Assert(command != null, "Caller should validate the parameter");
             _command = command;
+
+            _pipeline = ScriptBlock.Create("Out-String -Stream").GetSteppablePipeline();
+            _pipeline.Begin(true);
         }
 
         #endregion constructor
 
-        /// <summary>
-        /// Input is collected in this list
-        /// </summary>
-        private ArrayList _inputList = new ArrayList();
+        private SteppablePipeline _pipeline;
+
         /// <summary>
         /// Add an object to write to process
         /// </summary>
         /// <param name="input"></param>
         internal void Add(object input)
         {
-            _inputList.Add(input);
-        }
+            if (_stopping || _streamWriter == null)
+            {
+                // if _streamWriter is already null, then we already called Done()
+                // so we should just discard the input.
+                return;
+            }
 
-        /// <summary>
-        /// Count of object in inputlist
-        /// </summary>
-        internal int Count
-        {
-            get
+            Array formattedObjects = _pipeline.Process(input);
+            foreach (var item in formattedObjects)
             {
-                return _inputList.Count;
+                string line = PSObject.ToStringParser(_command.Context, item);
```

Comment:
It looks you are not differentiating `Text` input from `Xml` input, but you are differentiating them when processing the output of native command. Is this a oversight?